### DISCUSSION TITLE
SSD202D - how to get started for testing.

### DISCUSSION THREAD

**Discussion by brentpicasso** [linux-chenxing/linux-chenxing.org#30] [General]
I have a number of the SSD202D related items on the way:
*  Qty 10 IDO-SOM2D02 from wireless-tag on alibaba (with 512MB flash)
* One 7" Smart Display from Seeed with the SSD202D
* A USB programming dongle that was recommended to me by wireless-tag

![image](https://user-images.githubusercontent.com/3267533/116788683-dd849200-aa5f-11eb-961b-b3a7676248be.png)

I could use some instructions on:
* What kernel to use and how to apply the necessary patches, then what kernel options I should enable.
* Getting started with the correct uboot build
* How to flash these SOMs. I'm guessing they are shipped without a working uboot+linux image.

If I can get some assistance, I'm happy to contribute a from-the-ground-up how to get started with this chip so others can easily repeat the process. 

Thanks!

**Comment by fifteenhex** [comment]
Hi,

There is a little bit of info on how to get started here:
http://linux-chenxing.org/infinity2/ido-som2d01/

I can't remember if the modules come blank or not. They are fairly easy to flash though. You connect the flashing dongle to the pm_uart port and run the ISP tool.

I think I can put together a buildroot setup that builds everything you need if that would help?

**Comment by brentpicasso** [comment]
A buildroot setup would be fantastic.  Hardware arrives on Monday; excited!

**Reply by brentpicasso** [reply]
Note, the The ISP link to the ISP on the page you linked is 404.  http://linux-chenxing.org/infinity2/ido-som2d01/isp

**Reply by fifteenhex** [reply]
> A buildroot setup would be fantastic. Hardware arrives on Monday; excited!

I've started putting together a buildroot setup here:
https://github.com/fifteenhex/buildroot_idosom2d01

I haven't tested it yet but I will in the next few days. Once I know it works I'll a tag a release and there will be some binaries you can flash onto your modules with the ISP tool. That will save you having to work out how to build u-boot for now and give you a known working image before you start hacking.

I've fixed the link to the ISP page. I guess we really need to document how to use the windows ISP tool.
I think there are some public links to it now.

**Reply by brentpicasso** [reply]
Thanks, will check it out!  FWIW, my development environment is Linux (ubuntu).

**Reply by fifteenhex** [reply]
> my development environment is Linux (ubuntu).

ubuntu should work fine. You'll need a windows VM for the ISP tool though. I have a windows 7 VM and that seems to work fine.
Eventually I'll finish up a method of flashing straight from Linux.. ;)

**Comment by brentpicasso** [comment]
Side question, what kind of heat management does this need?  Does it need a heat sink if both cores are pegged at 100%?

**Reply by fifteenhex** [reply]
I haven't used a heat sink with the cores overclocked to 1.8GHz. I think if it's in an open area it's fine. If you put it into a case you might need a sink to make up for there being no air flow.

**Reply by fifteenhex** [reply]
Oh and I forgot:

There is an internal temp sensor that could be linked to cpufreq to throttle if things are getting warm.

**Comment by brentpicasso** [comment]
Ok, I have the 7" display, as well as the USB dongle they suggested I get.  The sparse documentation implies it provides an I2C interface. 

* Can this dongle work with the 7" display? It appears there's a debug port near the top.  
* Is this the ISP interface? 
* Is there a uart for getting the console? It appears the USB-C connection does not have a serial connection, just power. 
*
![image](https://user-images.githubusercontent.com/3267533/117396024-fea91080-aead-11eb-80ba-d0b0cfeefbd0.png)

**Reply by fifteenhex** [reply]
The debug interface is a UART and an I2C slave. The dongle you have should work for flashing.
For serial you just connect a USB->Serial to the same pins.

**Reply by kamomil** [reply]
Hi,
Do you know what are the rx, tx and grd in the debug socket (J12) and also what is the baudrate?
thanks

**Reply by kamomil** [reply]
oh, I see now it is already written on the board

**Reply by kamomil** [reply]
@brentpicasso hi, I have the same board.
I have UART working with the default software that comes with the board (OpenWrt).
I want to compile and flash my own OS (probably based on that buildroot published here)

Could you share the link for the USB dongle you have? Did you use it only for UART or also for the ISP burning?

**Reply by brentpicasso** [reply]
I got it from the wireless-tag company on alibaba.  They were able to supply the dongle for me.  I would like to know if there's a different way to flash these devices for the first time - I would not want to be tied to mystery proprietary hardware. @fifteenhex - any suggestions on it?

**Comment by brentpicasso** [comment]
I'm getting geared up to try your buildroot image.  What UART is it configured to use for the console?

**Reply by fifteenhex** [reply]
PM_UART is used for the console. Same connector as the debug tool.

**Comment by brentpicasso** [comment]
I got information on connecting the USB programming dongle to the evaluation board. 

What software is used for flashing bits to the SOM? Is it OpenOCD, or something else?

**Reply by fifteenhex** [reply]
You need the mstar/sigmastar isp tool. It should be in the SDK. Unfortunately it's a crappy windows tool but you only need to use it once.

**Reply by brentpicasso** [reply]
Thanks, I'll dig around - hopefully I have it there somewhere. :)

**Reply by brentpicasso** [reply]
I see the USB tool is an FTDI device when plugged into Linux. I'm guessing a Linux version of the ISP tool would involve emulating the serial protocol?

**Reply by brentpicasso** [reply]
Ah, I see from the Uboot instructions that the Flashrom tool needs to be updated to handle NAND flash.

**Reply by fifteenhex** [reply]
yeah, writing proper drivers for NAND isn't easy. I've been tinkering with making a dummy spi driver for the kernel so that we can have a userland tool that takes the spi commands for the flash and drives the programming interface on the chip.
Extracting the nand code from u-boot and making it a userland too might be better in the long run though.

**Reply by fifteenhex** [reply]
BTW,.. does your module have foresee flash? I noticed an issue with foresee flash and the current u-boot where the ubi partition gets corrupted.

**Reply by brentpicasso** [reply]
it does have Forsee flash, correct.  512MB variant.
